torch/_dynamo/variables/misc.py

```diff
@@ -389,55 +386,58 @@ def visit(node):
             # we install any guards accumulated from tracing. If not, we graph break. We trace backward, and evaluate
             # for soundness, same as forward, except with more strictness. We enable a strict mode on the tx, and
             # reject certain ops when running under this strict mode. If the backward trace is sound, we discard the
-            # trace by restoring. Otherwise, we raise.
+            # trace by restarting analysis and skipping the soundness evalution on the subsequent run. Otherwise, we raise.
 
             # if both the forward and backward traces are sound, we write the autograd function’s apply into the graph.
 
             # For tracing forward and backward, we use UserFunctionVariable. Although it does not directly contribute
             # to soundness evaluation, it plus a  GlobalSource makes sure we can produce valid guards,
             # and that we can inline properly here. Inlining is required in order to be able to ensure that the
             # soundness evaluation works as described above.
-            graph_checkpoint, checkpoint = tx.output.graph, tx.copy_graphstate()
+            speculation = tx.speculate_autograd_fn()
+
+            if not speculation.succeeded:
+                # Don't accumulate speculation history
+                # because we will not take this path if it succeeds
+                # causing a divergence if we ever return here from other restarts
```

Comment:
This isn't the clearest to me - maybe write a note and reference it?

Reply:
I put a note inline